File: src/main/java/net/ndrei/teslacorelib/gui/SideConfigSelector.java
package net.ndrei.teslacorelib.gui;

import com.google.common.collect.Lists;
import net.ndrei.teslacorelib.capabilities.inventory.SidedItemHandlerConfig;
import net.ndrei.teslacorelib.containers.BasicTeslaContainer;
import net.ndrei.teslacorelib.inventory.BoundingRectangle;
import net.ndrei.teslacorelib.inventory.ColoredItemHandlerInfo;

import java.util.List;

/**
 * Created by CF on 2016-12-22.
 */
public class SideConfigSelector extends BasicContainerGuiPiece {
    private SidedItemHandlerConfig sidedConfig;
    private SideConfigurator configurator;
    private int selectedInventory = -1;

    public SideConfigSelector(int left, int top, int width, int height, SidedItemHandlerConfig sidedConfig, SideConfigurator configurator) {
        super(left, top, width, height);

        this.sidedConfig = sidedConfig;
        this.configurator = configurator;
    }

    @Override
    public void drawBackgroundLayer(BasicTeslaGuiContainer container, int guiX, int guiY, float partialTicks, int mouseX, int mouseY) {
        List<ColoredItemHandlerInfo> colors = this.sidedConfig.getColoredInfo();
        if ((colors != null) && (colors.size() > 0)) {
            container.bindDefaultTexture();
            for (int i = 0; (i < colors.size()) && (i < 9); i++) {
                container.drawTexturedRect(
                        this.getLeft() + 2 + i * 18, this.getTop() + 2,
                        (i == this.selectedInventory) ? 128 : 110, 210, 14, 14);
                container.drawFilledRect(
                        guiX + this.getLeft() + 4 + i * 18, guiY + this.getTop() + 4,
                        10, 10, 0xFF000000 + colors.get(i).getColor().getMapColor().colorValue);
            }
        }
    }

    @Override
    public void drawForegroundLayer(BasicTeslaGuiContainer container, int guiX, int guiY, int mouseX, int mouseY) {
        List<ColoredItemHandlerInfo> colors = this.sidedConfig.getColoredInfo();

        if (this.isInside(container, mouseX, mouseY)) {
            int localY = mouseY - guiY - this.getTop();
            if ((localY >= 2) && (localY <= 14)) {
                int localX = mouseX - guiX - this.getLeft();
                int index = localX / 18;
                if ((index != this.selectedInventory) && (index >= 0) && (index < colors.size()) && (colors.get(index).getHighlight() != null)) {
                    localX = localX - index * 18;
                    if ((localX >= 2) && (localX <= 14)) {
                        BoundingRectangle box = colors.get(index).getHighlight();
                        container.drawFilledRect(box.getLeft(), box.getTop(), box.getWidth(), box.getHeight(),
                                0x42000000 + colors.get(index).getColor().getMapColor().colorValue,
                                0xFF000000 + colors.get(index).getColor().getMapColor().colorValue);
                    }
                }
            }
        }

        if ((this.selectedInventory >= 0) && (this.selectedInventory < colors.size())) {
            BoundingRectangle box = colors.get(this.selectedInventory).getHighlight();
            container.drawFilledRect(box.getLeft(), box.getTop(), box.getWidth(), box.getHeight(),
                    0x42000000 + colors.get(this.selectedInventory).getColor().getMapColor().colorValue,
                    0xFF000000 + colors.get(this.selectedInventory).getColor().getMapColor().colorValue);
        }
    }

    @Override
    public void drawForegroundTopLayer(BasicTeslaGuiContainer container, int guiX, int guiY, int mouseX, int mouseY) {
        if (this.isInside(container, mouseX, mouseY)) {
            int localY = mouseY - guiY - this.getTop();
            if ((localY >= 2) && (localY <= 14)) {
                int localX = mouseX - guiX - this.getLeft();
                int index = localX / 18;
                List<ColoredItemHandlerInfo> colors = this.sidedConfig.getColoredInfo();
                if ((index >= 0) && (index < colors.size()) && (colors.get(index).getHighlight() != null)) {
                    localX = localX - index * 18;
                    if ((localX >= 2) && (localX <= 14)) {
                        String label = colors.get(index).getName();
                        if ((label != null) && (label.length() > 0)) {
                            container.drawTooltip(Lists.newArrayList(label),
                                    this.getLeft() + (index * 18) + 9,
                                    this.getTop() + this.getHeight() / 2);
                        }
                    }
                }
            }
        }
    }

    @Override
    public void mouseClicked(BasicTeslaGuiContainer container, int mouseX, int mouseY, int mouseButton) {
        int oldIndex = this.selectedInventory;
        this.selectedInventory = -1;
        if (this.isInside(container, mouseX, mouseY)) {
            int localY = mouseY - container.getGuiTop() - this.getTop();
            if ((localY >= 2) && (localY <= 14)) {
                int localX = mouseX - container.getGuiLeft() - this.getLeft();
                int index = localX / 18;
                List<ColoredItemHandlerInfo> colors = this.sidedConfig.getColoredInfo();
                if ((index != oldIndex) && (index >= 0) && (index < colors.size()) && (colors.get(index).getHighlight() != null)) {
                    localX = localX - index * 18;
                    if ((localX >= 2) && (localX <= 14)) {
                        this.selectedInventory = index;
                    }
                }
            }
        }

        if (this.configurator != null) {
            this.configurator.setSelectedInventory(this.selectedInventory);
        }

        BasicTeslaContainer slots = container.getTeslaContainer();
        if (slots != null) {
            if (this.selectedInventory >= 0) {
                slots.hidePlayerInventory();
            }
            else {
                slots.showPlayerInventory();
            }
        }
    }
}

File: src/main/kotlin/net/ndrei/teslacorelib/gui/SideConfigSelector.kt
package net.ndrei.teslacorelib.gui

import com.google.common.collect.Lists
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import net.ndrei.teslacorelib.capabilities.inventory.SidedItemHandlerConfig

/**
 * Created by CF on 2017-06-28.
 */
class SideConfigSelector(left: Int, top: Int, width: Int, height: Int, private val sidedConfig: SidedItemHandlerConfig, private val configurator: SideConfigurator?)
    : BasicContainerGuiPiece(left, top, width, height) {
    private var selectedInventory = -1

    @SideOnly(Side.CLIENT)
    override fun drawBackgroundLayer(container: BasicTeslaGuiContainer<*>, guiX: Int, guiY: Int, partialTicks: Float, mouseX: Int, mouseY: Int) {
        val colors = this.sidedConfig.coloredInfo
                .take(8) // to make sure there is no overflow
        if (colors.isNotEmpty()) {
            container.bindDefaultTexture()
            var i = 0
            colors.forEach {
                container.drawTexturedRect(
                        this.left + 2 + i * 18, this.top + 2,
                        if (i == this.selectedInventory) 128 else 110, 210, 14, 14)
                container.drawFilledRect(
                        guiX + this.left + 4 + i * 18, guiY + this.top + 4,
                        10, 10, 0xFF000000.toInt() + it.color.colorValue)
                i++
            }
        }
    }

    @SideOnly(Side.CLIENT)
    override fun drawForegroundLayer(container: BasicTeslaGuiContainer<*>, guiX: Int, guiY: Int, mouseX: Int, mouseY: Int) {
        val colors = this.sidedConfig.coloredInfo

        if (this.isInside(container, mouseX, mouseY)) {
            val localY = mouseY - guiY - this.top
            if (localY in 2..14) {
                var localX = mouseX - guiX - this.left
                val index = localX / 18
                if (index != this.selectedInventory && index >= 0 && index < colors.size && index < 8 && colors[index].highlight != null) {
                    localX = localX - index * 18
                    if (localX in 2..14) {
                        val box = colors[index].highlight
                        container.drawFilledRect(box.left, box.top, box.width, box.height,
                                0x42000000 + colors[index].color.colorValue,
                                0xFF000000.toInt() + colors[index].color.colorValue)
                    }
                }
            }
        }

        if (this.selectedInventory >= 0 && this.selectedInventory < colors.size) {
            val box = colors[this.selectedInventory].highlight
            container.drawFilledRect(box.left, box.top, box.width, box.height,
                    0x42000000 + colors[this.selectedInventory].color.colorValue,
                    0xFF000000.toInt() + colors[this.selectedInventory].color.colorValue)
        }
    }

    @SideOnly(Side.CLIENT)
    override fun drawForegroundTopLayer(container: BasicTeslaGuiContainer<*>, guiX: Int, guiY: Int, mouseX: Int, mouseY: Int) {
        if (this.isInside(container, mouseX, mouseY)) {
            val localY = mouseY - guiY - this.top
            if (localY in 2..14) {
                var localX = mouseX - guiX - this.left
                val index = localX / 18
                val colors = this.sidedConfig.coloredInfo
                if (index >= 0 && index < colors.size && index < 8) {
                    localX -= index * 18
                    if (localX in 2..14) {
                        val label = colors[index].name
                        if (label.isNotEmpty()) {
                            container.drawTooltip(Lists.newArrayList(label),
                                    this.left + index * 18 + 9,
                                    this.top + this.height / 2)
                        }
                    }
                }
            }
        }
    }

    @SideOnly(Side.CLIENT)
    override fun mouseClicked(container: BasicTeslaGuiContainer<*>, mouseX: Int, mouseY: Int, mouseButton: Int) {
        val oldIndex = this.selectedInventory
        this.selectedInventory = -1
        if (this.isInside(container, mouseX, mouseY)) {
            val localY = mouseY - container.guiTop - this.top
            if (localY in 2..14) {
                var localX = mouseX - container.guiLeft - this.left
                val index = localX / 18
                val colors = this.sidedConfig.coloredInfo
                if (index != oldIndex && index >= 0 && index < colors.size && index < 8 && colors[index].highlight != null) {
                    localX -= index * 18
                    if (localX in 2..14) {
                        this.selectedInventory = index
                    }
                }
            }
        }

        if (this.configurator != null) {
            this.configurator.setSelectedInventory(this.selectedInventory)
        }

        val slots = container.teslaContainer
        if (slots != null) {
            if (this.selectedInventory >= 0) {
                slots.hidePlayerInventory()
            } else {
                slots.showPlayerInventory()
            }
        }
    }
}
